Sources/SwiftFusion/Core/FixedSizeMatrix.swift

```diff
@@ -43,65 +47,312 @@ public struct FixedSizeMatrix<Rows: Equatable & FixedSizeArray> where Rows.Eleme
+  @usableFromInline
+  func vjpDot(_ other: Self) -> (value: Double, pullback: (Double) -> (Self, Self)) {
+    return (self.dot(other), { v in (v * other, v * self) })
+  }
+
+  public static var dimension: Int {
+    return shape[0] * shape[1]
+  }
+
+  public static var standardBasis: [Self] {
+    (0..<dimension).map { i in
+      var v = zero
+      v[i] = 1
+      return v
+    }
   }
 }
 
+// MARK: - Matrix math.
+
 extension FixedSizeMatrix {
-  public static func * (_ lhs: Double, _ rhs: Self) -> Self {
-    var result = rhs
-    for i in 0..<Self.rowCount {
-      for j in 0..<Self.columnCount {
-        result[i, j] *= lhs
+  public static var isSquare: Bool {
+    return shape[0] == shape[1]
+  }
+
+  /// Creates the outer product of `lhs` and `rhs`.
+  ///
+  /// - Requires: `Self.isSquare`.
+  public init(outerProduct lhs: Rows.Element, _ rhs: Rows.Element) {
+    precondition(Self.isSquare, "init(outerProduct:_:) requires a square matrix")
+    self = Self.zero
+    for i in 0..<Self.shape[0] {
+      for j in 0..<Self.shape[0] {
+        self[i, j] = lhs[i] * rhs[j]
+      }
+    }
+  }
+
+  /// Returns the transpose of `self`.
+  ///
+  /// - Requires: `Self.isSquare`.
+  @differentiable
+  public func transposed() -> Self {
+    precondition(Self.isSquare, "transposed() requires a square matrix")
+    var r = Self.zero
+    for i in 0..<Self.shape[0] {
+      for j in 0..<Self.shape[0] {
+        r[i, j] = self[j, i]
       }
     }
-    return result
+    return r
+  }
+
+  @derivative(of: transposed)
+  @usableFromInline
+  func vjpTransposed() -> (value: Self, pullback: (Self) -> Self) {
+    return (transposed(), { $0.transposed() })
+  }
+}
+
+/// Returns the matrix-vector product of `lhs` and `rhs`.
+///
+/// Note: This is currently only implemented for square matrices, but could be extended later.
+///
+/// - Requires: `type(of: lhs).isSquare`.
+@differentiable
+public func matvec<Rows>(_ lhs: FixedSizeMatrix<Rows>, _ rhs: Rows.Element) -> Rows.Element {
+  precondition(type(of: lhs).isSquare, "matvec only implemented for square matrices")
+  var r = Rows.Element.zero
+  for i in 0..<Rows.Element.dimension {
+    for j in 0..<Rows.Element.dimension {
+      r[i] += lhs[i, j] * rhs[j]
+    }
+  }
+  return r
+}
+
+@derivative(of: matvec)
+@usableFromInline
+func vjpMatvec<Rows>(_ lhs: FixedSizeMatrix<Rows>, _ rhs: Rows.Element)
+  -> (value: Rows.Element, pullback: (Rows.Element) -> (FixedSizeMatrix<Rows>, Rows.Element))
+{
+  return (
+    matvec(lhs, rhs),
+    { v in
+      (FixedSizeMatrix<Rows>(outerProduct: v, rhs), matvec(lhs.transposed(), v))
+    }
+  )
+}
+
+/// Returns the matrix-matrix product of `lhs` and `rhs`.
+///
+/// Note: This is currently only implemented for the product of square matrices of identical shape,
+/// but could be extended later.
+///
+/// - Requires: `type(of: lhs).isSquare`.
+@differentiable
+public func matmul<Rows>(_ lhs: FixedSizeMatrix<Rows>, _ rhs: FixedSizeMatrix<Rows>)
```

Comment:
A naive impl of matmul would be just repeat matvec for n times, which would eliminate this code estate here? Feel free to ignore :)

Reply:
It's not easy to do it this way yet, because we don't have a function for forming the result matrix from column vectors. And the type of that function is nontrivial, so it might be a bit hard to add. But this definitely seems like a good thing to do after the method exists. (I imagine that we'll figure out how to add it eventually because it'll be very convenient for lots of things).